core/src/main/java/hudson/model/Items.java

```diff
@@ -440,6 +470,121 @@ String name(Item i) {
 
+    private static class AllItemsIterable<T extends Item> implements Iterable<T> {
+
+        /**
+         * The authentication we are iterating as.
+         */
+        private final Authentication authentication;
+        /**
+         * The root we are iterating from.
+         */
+        private final ItemGroup root;
+        /**
+         * The type of item we want to return.
+         */
+        private final Class<T> type;
+
+        private AllItemsIterable(ItemGroup root, Authentication authentication, Class<T> type) {
+            this.root = root;
+            this.authentication = authentication;
+            this.type = type;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public Iterator<T> iterator() {
+            return new AllItemsIterator();
+        }
+
+        private class AllItemsIterator implements Iterator<T> {
+
+            /**
+             * The stack of {@link ItemGroup}s that we have left to descend into.
+             */
+            private final Stack<ItemGroup> stack = new Stack<>();
+            /**
+             * The iterator of the current {@link ItemGroup} we
+             */
+            private Iterator<Item> delegate = null;
+            /**
+             * The next item.
+             */
+            private T next = null;
+
+            private AllItemsIterator() {
+                // put on the stack so that hasNext() is the only place that has to worry about authentication
+                // alternative would be to impersonate and populate delegate.
+                stack.push(root);
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public boolean hasNext() {
+                if (next != null) {
+                    return true;
+                }
+                while (true) {
+                    if (delegate == null || !delegate.hasNext()) {
+                        if (stack.isEmpty()) {
+                            return false;
+                        }
+                        ItemGroup group = stack.pop();
+                        // group.getItems() is responsible for performing the permission check so we will not repeat it
+                        if (Jenkins.getAuthentication() == authentication) {
+                            delegate = group.getItems().iterator();
+                        } else {
+                            // slower path because the caller has switched authentication
+                            // we need to keep the original authentication so that allItems() can be used
+                            // like getAllItems() without the cost of building the entire list up front
+                            try (ACLContext ctx = ACL.as(authentication)) {
```

Comment:
For 99.9% of use cases, people will not be changing authentication while iterating, and hence

```
if (Jenkins.getAuthentication() == authentication) {
    delegate = group.getItems().iterator();
}
```

Will be the path taken... which will just be inlined to a call to `SecurityContextHolder.getContext().getAuthentication()`... given that the `getItems()` will be making this call anyway, the only side-effect in terms of performance is that we will have pre-seeded the CPU cache with the object chain for the `SecurityContextHolder.getContext().getAuthentication()` call anyway.

There are some cases where the authentication used to enumerate the items does not match to the current authentication. 

In Core there is one example: https://github.com/jenkinsci/jenkins/pull/2665/files#diff-46dd855f38f94f25c68cbdae7e6ba64bL243 the example is correct:

* User decides to rename / move a job
* The operation runs as the user's permission
* The item's `renameTo` does escalation to ACL.SYSTEM for the parts that require this escalation, but otherwise is running as the user who did the rename.
* Thus, while iterating the iterable, the items should be the items visible to ACL.SYSTEM but the current authentication is not ACL.SYSTEM...

The second half of the branch is to cover this exact case. 

HTH